lib/phoenix_html/link.ex

```diff
@@ -165,4 +175,30 @@ defmodule Phoenix.HTML.Link do
 
     {value, opts}
   end
+
+  defp valid_destination!(to, calling_func) do
+    if invalid_destination?(to) do
+      raise ArgumentError, """
+      unsupported scheme given to #{calling_func}. In case you want to link to an
+      unknown or unsafe scheme, such as javascript, use a tuple: {:javascript, rest}
+      """
+    end
+
+    case to do
+      {:safe, _} = dest -> dest
+      {scheme, dest} -> "#{scheme}:#{dest}"
+      dest -> dest
+    end
+  end
+
+  for scheme <- @valid_uri_schemes do
+    defp invalid_destination?(unquote(scheme) <> _), do: false
+  end
+  defp invalid_destination?({scheme, to}) when is_atom(scheme) do
+    invalid_destination?(to)
```

Comment:
Now I think that this code is not going to allow this: `{:javascript, "console.log({one: 1})"}` because it still checks for `:`. We also need to make sure this works: `{:javascript, {:safe, ...}}`.

Since there are so many corner cases, it may be better to do it in two steps instead of making it recursive so we also avoid `{:safe, {:javascript, ...}}` and other combinations. Maybe this:

```elixir
defp valid_destination!({:safe, to}) do
  {:safe, valid_destination_string!(to)}
end
defp valid_destination!({other, to}) do
  [other, ?:, to]
end
defp valid_destination!(to) do
  valid_destination_string!(to)
end
```

Reply:
PS: I slightly edited the code above.

Reply:
> Now I think that this code is not going to allow this: {:javascript, "console.log({one: 1})"}

Oops, you're right. Maybe I should stick to security! 😅